File: ase-backend/src/main/java/de/corvinrose/ase/controller/LikeController.java
package de.corvinrose.ase.controller;

import de.corvinrose.ase.model.Like;
import de.corvinrose.ase.service.LikeService;
import java.util.List;
import java.util.UUID;
import lombok.AllArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

@AllArgsConstructor
@RestController
@RequestMapping("/likes")
public class LikeController {

    private LikeService likeService;

    @GetMapping
    public ResponseEntity<List<Like>> getAllLikes() {
        List<Like> users = likeService.findAllLikes();
        return new ResponseEntity<>(users, HttpStatus.OK);
    }

    @GetMapping("/shader/{id}")
    public ResponseEntity<List<UUID>> getAllLikesByShaderId(@PathVariable("id") UUID id) {
        List<UUID> userIds = likeService.findAllLikesByShader(id);
        return new ResponseEntity<>(userIds, HttpStatus.OK);
    }

    @GetMapping("/user/{id}")
    public ResponseEntity<List<UUID>> getAllLikesByUserId(@PathVariable("id") UUID id) {
        List<UUID> shaderIds = likeService.findAllLikesByUser(id);
        return new ResponseEntity<>(shaderIds, HttpStatus.OK);
    }

    @PostMapping("/add")
    public ResponseEntity<Like> addLike(@RequestBody Like like) {
        Like newLike = likeService.addLike(like);
        return new ResponseEntity<>(newLike, HttpStatus.CREATED);
    }

    @Transactional
    @PostMapping("/delete")
    public ResponseEntity<Like> deleteLike(@RequestBody Like like) {
        likeService.deleteLike(like);
        return new ResponseEntity<>(HttpStatus.OK);
    }
}

File: backend/Port/Web/Controller/LikeController.php
<?php

namespace Port\Web\Controller;

use Core\App\Provider\BaseController;
use Core\App\Service\AuthService;
use Core\App\Service\LikeService;
use Core\Domain\Model\Routing\HttpStatus;
use Core\Domain\Model\Routing\ResponseEntity;
use Core\Domain\Model\LikeModel;
use Exception;

class LikeController extends BaseController {
    private LikeService $likeService;
    private AuthService $authService;

    public function __construct() {
        $this->likeService = new LikeService();
        $this->authService = new AuthService();
    }

    /**
     * @Get /v1/likes
     * @throws Exception
     */
    public function getAllLikes(): ResponseEntity {
        return new ResponseEntity($this->likeService->findAllLikes(), HttpStatus::OK());
    }

    /**
     * @Get /v1/likes/shader/{id}
     * @throws Exception
     */
    public function getAllLikesByShaderId(string $id): ResponseEntity {
        return new ResponseEntity($this->likeService->findAllLikesByShader($id), HttpStatus::OK());
    }

    /**
     * @Get /v1/likes/user/{id}
     * @throws Exception
     */
    public function getAllLikesByUserId(string $id): ResponseEntity {
        return new ResponseEntity($this->likeService->findAllLikesByUser($id), HttpStatus::OK());
    }

    /**
     * @Post /v1/likes/add
     * @throws Exception
     */
    public function addLike(): ResponseEntity {
        $requestBody = file_get_contents("php://input");
        $like = LikeModel::fromJsonObject(json_decode($requestBody));

        $authStatus = $this->authService->authStatus();
        $user = $this->authService->findUserByEmail($authStatus->userName);
        if ($user->id !== $like->userId) {
            return $this->authService->unauthorized();
        }
        return new ResponseEntity($this->likeService->addLike($like), HttpStatus::CREATED());
    }

    /**
     * @Post /v1/likes/delete
     * @throws Exception
     */
    public function deleteLike(): ResponseEntity {
        $requestBody = file_get_contents("php://input");
        $like = LikeModel::fromJsonObject(json_decode($requestBody));

        $authStatus = $this->authService->authStatus();
        $user = $this->authService->findUserByEmail($authStatus->userName);
        if ($user->id !== $like->userId) {
            return $this->authService->unauthorized();
        }
        return new ResponseEntity($this->likeService->deleteLike($like), HttpStatus::OK());
    }
}
